### ISSUE TITLE
CharacterBody3D sticks, stutters, slows down when sliding against walls

### ISSUE THREAD

**Issue by tom-jk** [godot-jolt/godot-jolt#529]
godot 4.1.1
godot-jolt 0.5.0-rc1

In my 3d platformer game the character sticks to, stutters on and slows down against walls.

I have attached a project that I hope will be useful for testing this.

[joltWallSlideTest0.zip](https://github.com/godot-jolt/godot-jolt/files/12244228/joltWallSlideTest0.zip)

Video (sped-up to trade off length versus file size):

[joltwallslidetest0demo.webm](https://github.com/godot-jolt/godot-jolt/assets/64640811/70541bc0-827e-4f6f-ba97-3facc0f00592)

When you press run, a thousand-odd characterbody3d's with box collision will walk across a section of trimesh level collision at some amount of slope into a wall at some angle. You can also click one for a closer look (click again to return). The project will run slow but I don't think (?) this will affect results. If desired I could add a smaller, more focused test scene.

Generally, characters should slide along walls until they fall off the level (except those on green sections with a flat wall). You'll see under a lot of typical configurations that some will get stuck. If you move them by repeatedly tapping up, most (but usually not all) will make progress along their wall. The most success appears when block on walls is disabled, but normally you'll want it enabled.

note: haven't tested under godot physics yet, but I expect it might be similar (I think I've seen some allusions to similar issues being more to do with move_and_slide than the physics itself?)

**Comment by mihe** [comment]
That's quite the setup! Thank you for taking the time to report the issue.

It does seem like Godot Jolt performs worse than Godot Physics on some of these, even when using the more traditional capsule shape for the player. Specifically the left side of the lower half seems to be more adversely affected than the rest, as your video clearly shows already. Godot Physics is not entirely immune to this though, and there are some stragglers here and there.

Similarly, when just adding a third-person controller and walking around on these platforms, the same relationship can be observed, where with Godot Jolt you might end up getting stuck a bit more often, specifically on the walls of some of the steeper platforms, and with Godot Physics you might not get stuck as often, but you do eventually.

I suspect this will end up being similar to #524, where it's likely that some flaw in the implementation of `move_and_slide` has been made more prominent in Godot Jolt due to some differences in the types of collision detection that the two engines use.

I will need to step through the code to see what's actually going on here though.

**Comment by EIREXE** [comment]
> That's quite the setup! Thank you for taking the time to report the issue.
> 
> It does seem like Godot Jolt performs worse than Godot Physics on some of these, even when using the more traditional capsule shape for the player. Specifically the left side of the lower half seems to be more adversely affected than the rest, as your video clearly shows already. Godot Physics is not entirely immune to this though, and there are some stragglers here and there.
> 
> Similarly, when just adding a third-person controller and walking around on these platforms, the same relationship can be observed, where with Godot Jolt you might end up getting stuck a bit more often, specifically on the walls of some of the steeper platforms, and with Godot Physics you might not get stuck as often, but you do eventually.
> 
> I suspect this will end up being similar to #524, where it's likely that some flaw in the implementation of `move_and_slide` (which is largely implemented on the Godot side of things) has been made more prominent in Godot Jolt due to some differences in the types of collision detection that the two engines use.
> 
> I will need to step through the code to see what's actually going on here though.

Slightly related question, I assume with godot-jolt we are using Godot's implementation of CharacterBody3D, would it be possible to expose the jolt native character controller?

**Comment by mihe** [comment]
> I assume with godot-jolt we are using Godot's implementation of CharacterBody3D

Yes, `move_and_slide` is [mostly](https://github.com/godotengine/godot/blob/ff5c884153d1c15bde3ee87aa295e52bec3a6340/scene/3d/physics_body_3d.cpp#L1249-L1509) implemented on the Godot side of things, but the actual [collision checks](https://github.com/godotengine/godot/blob/ff5c884153d1c15bde3ee87aa295e52bec3a6340/scene/3d/physics_body_3d.cpp#L116) it relies on is delegated to the physics server, so this extension still plays some part in it.

> would it be possible to expose the jolt native character controller?

Maybe. I haven't really looked into it. I've tried to limit the amount of custom nodes I expose, since nodes exposed from GDExtension can't be interact with programmatically from C#, which is very unfortunate, especially from something as logic-heavy as a character controller. I'm also not sure how exactly the inheritance of `PhysicsBody3D` would play out in practice, which it would have to inherit from in order to work with things like `Area3D` and whatnot.

**Comment by mihe** [comment]
This problem should be fixed once #559 is merged, and seems to perform slightly better than Godot Physics even.

Thanks a lot for providing this test scene, @tom-jk, it was incredibly useful for tracking this one down. It also helped me find and fix another bug related to `move_and_slide`. It honestly made me regret not having spent more time creating some sort of test suite, like I promised myself I would once upon a time.

Here's what it looks like after a default run when using Jolt, before the changes:

![Jolt Before](https://github.com/godot-jolt/godot-jolt/assets/4884246/78b99e20-7fca-4f21-b1ad-781c175a28b4)

Here's what it looks like after a default run when using Jolt, after the changes:

![Jolt After](https://github.com/godot-jolt/godot-jolt/assets/4884246/a903a481-186b-4890-9c67-6281eab153ba)

Here's what it looks like after a default run when using Godot Physics:

![Godot](https://github.com/godot-jolt/godot-jolt/assets/4884246/d3fbf7d5-ea86-4607-a72f-53f6de3cd60e)

**Comment by mihe** [comment]
Note that it still stutters slightly when sliding along the walls though, but from what I can see that's pretty much identical to how Godot Physics behaves, so I'm assuming that's due to how `move_and_slide` is implemented on the Godot side of things.

**Comment by mihe** [comment]
The fix for this is included in the newly released [0.7.0-stable](https://github.com/godot-jolt/godot-jolt/releases/tag/v0.7.0-stable).

**Comment by tom-jk** [comment]
Thanks, that's definitely a big improvement:

![Screenshot from 2023-08-30 11-08-57](https://github.com/godot-jolt/godot-jolt/assets/64640811/100fb913-3e42-4dc0-a10f-c739f97d2deb)

It differs a little from yours above, but other commits have happened in the meantime so that's not too unexpected.

I've given it a brief test run in my main project but little seems to have changed there, so I'll have to look closer at what it is I might be doing wrong - though I, too also expect that it now probably lies more with godot than jolt.

**Comment by mihe** [comment]
> I've given it a brief test run in my main project but little seems to have changed there

If you find that it behaves better when using Godot Physics then I'd be interested in seeing a more representative repro. It doesn't have to be as flashy as the one you provided here.

**Comment by tom-jk** [comment]
Best I've managed so far is this where both get stuck, but GP gets a little further up the slope and maybe does it a little smoother?

[joltwallslidetest1.webm](https://github.com/godot-jolt/godot-jolt/assets/64640811/8e8cf787-a6ad-4549-8529-c022126caa3e)

GP looks to snag on a spurious wall-like normal, Jolt's not so clear.

[joltWallSlideTest1_share.zip](https://github.com/godot-jolt/godot-jolt/files/12597953/joltWallSlideTest1_share.zip)

Pressing Enter while the player is stuck will print something like the following:

```
(Godot Physics)

Slide Collision Count = 2
  slide #0: depth = 0.00106543221045, travel = (-0.001497, 0, -0.001663), remainder = (-0.026376, 0, -0.029307), collision count = 3
    collision #0: pos = (0.639648, 8.602197, -12.28522), nor = (0, 0.819152, 0.573576)
    collision #1: pos = (-0.360352, 8.602197, -12.28522), nor = (0, 0.819152, 0.573576)
    collision #2: pos = (-0.360424, 8.602134, -11.28524), nor = (0.975253, 0.126812, -0.181106)
  slide #1: depth = 0.00189882400446, travel = (-0.000558, 0, -0.003003), remainder = (-0.005931, 0, -0.031939), collision count = 3
    collision #0: pos = (0.640631, 8.60288, -12.2862), nor = (0, 0.819152, 0.573576)
    collision #1: pos = (-0.359369, 8.60288, -12.2862), nor = (0, 0.819152, 0.573576)
    collision #2: pos = (0.640631, 8.60364, -12.28728), nor = (0, 0, 1)

(Jolt Physics)
Slide Collision Count = 2
  slide #0: depth = 0.00138680811506, travel = (-0.001089, 0, -0.00121), remainder = (-0.026785, 0, -0.029761), collision count = 4
    collision #0: pos = (-0.093299, 8.255125, -10.08977), nor = (0.975253, 0.126812, -0.181107)
    collision #1: pos = (-0.093299, 7.977794, -10.28395), nor = (1, 0, 0)
    collision #2: pos = (0.518514, 7.664889, -10.9466), nor = (0, 0.819152, 0.573576)
    collision #3: pos = (0.105318, 7.664889, -10.9466), nor = (0, 0.819152, 0.573576)
  slide #1: depth = 0.00115310552064, travel = (-0.000253, 0, -0.001365), remainder = (-0.006235, 0, -0.033577), collision count = 4
    collision #0: pos = (0.105429, 7.665005, -10.94677), nor = (0, 0.819152, 0.573576)
    collision #1: pos = (0.867551, 7.665005, -10.94677), nor = (0, 0.819152, 0.573576)
    collision #2: pos = (-0.093299, 8.234673, -10.10409), nor = (0.975253, 0.126812, -0.181107)
    collision #3: pos = (-0.093299, 8.255015, -10.08984), nor = (1, 0, 0)
```

![joltwallslidetest1](https://github.com/godot-jolt/godot-jolt/assets/64640811/8a402e4d-7e45-4fb7-bd17-2c07d31faefa)

---
I'll try to come up with something a little more definitive...

**Comment by mihe** [comment]
The getting stuck part was resolved recently in #581, but I haven't published an actual release with it yet.

The smoothness part of it is a more complicated topic, with no good solution, unfortunately. It's a side effect of the mechanism that fixes the getting stuck part. There are ways to make it better in this particular test case, by making [this](https://github.com/godot-jolt/godot-jolt/blob/afbcae0ebfb4e14bfd5bb6c8c17b9f38a98b7269/src/shapes/jolt_concave_polygon_shape_impl_3d.cpp#L7) configurable and setting it to a lower value, but that will simply lead to getting stuck in other places instead, as well as negatively affecting `RigidBody3D`. Frankly, this is probably one of those cases where the best option is to fix (or at least improve) `move_and_slide` from the Godot side of things instead, but that's a big task that I can't really afford to spend time on right now.

In general though you will see better (albeit not perfect) results by simply not using a `BoxShape3D` for your character and instead something like the more traditional `CapsuleShape3D`, so if that's an option for you then you might want to explore that.

**Comment by mihe** [comment]
Apologies for the long wait, but the fix for the getting stuck part of your latest repro is included in the newly released [0.8.0-rc1](https://github.com/godot-jolt/godot-jolt/releases/tag/v0.8.0-rc1). I'd appreciate it if you could give it a try and report back how well it works for you.

**Comment by tom-jk** [comment]
Haven't tested it properly otherwise yet, but I can affirm that it does solve getting stuck in that repro.

**Comment by mihe** [comment]
The fix for this is now also included in the newly released [0.8.0-stable](https://github.com/godot-jolt/godot-jolt/releases/tag/v0.8.0-stable).
